File: src/i1tbfssymmetric/Python/Queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def size(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def peek(self):
        if self.is_empty():
            print("QueueEmptyException")
            return None
        return self.head.value

    def enqueue(self, value):
        temp = Node(value)
        if self.head is None:
            self.head = temp
            self.tail = temp
        else:
            self.tail.next = temp
            self.tail = temp
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            print("QueueEmptyException")
            return None
        value = self.head.value
        self.head = self.head.next
        self.size -= 1
        return value

    def print_queue(self):
        if self.size == 0:
            return "[]"
        temp = self.head
        out = "["
        for _ in range(self.size):
            out += str(temp.value) + ", "
            temp = temp.next
        out = out[:-2]
        out += "]"
        return out

File: src/i1tbfssymmetric/TS/Queue.ts
class Node<T> {
    value: T;
    next: Node<T> | null;

    constructor(value: T) {
        this.value = value;
        this.next = null;
    }
}

export class Queue<T> {
    private head: Node<T> | null;
    private tail: Node<T> | null;
    private size: number;

    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public getSize(): number {
        return this.size;
    }

    public isEmpty(): boolean {
        return this.size === 0;
    }

    public peek(): T | null {
        if (this.isEmpty()) {
            console.log("QueueEmptyException");
            return null;
        }
        return this.head!.value;
    }

    public enqueue(value: T): void {
        const temp = new Node(value);
        if (this.head === null) {
            this.head = temp;
            this.tail = temp;
        } else {
            this.tail!.next = temp;
            this.tail = temp;
        }
        this.size++;
    }

    public dequeue(): T | null {
        if (this.isEmpty()) {
            console.log("QueueEmptyException");
            return null;
        }
        const value = this.head!.value;
        this.head = this.head!.next;
        this.size--;
        return value;
    }

    public printQueue(): string {
        if (this.size === 0) {
            return "[]";
        }
        let temp = this.head;
        let out = "[";
        while (temp) {
            out += `${temp.value}, `;
            temp = temp.next;
        }
        out = out.slice(0, -2) + "]";
        return out;
    }
}
